cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java

```diff
@@ -824,52 +825,83 @@ public KafkaBrokerConfigurationBuilder withUserConfiguration(KafkaConfiguration
+        printYammerMetricsReporters(userConfig, injectStrimziMetricsReporter);
 
-        // print user config with Strimzi injections
         if (!userConfig.getConfiguration().isEmpty()) {
-            printSectionHeader("User provided configuration with Strimzi injections");
+            printSectionHeader("User provided configuration");
             writer.println(userConfig.getConfiguration());
             writer.println();
         }
 
         return this;
     }
 
-    /**
-     * Adds the Kafka metric.reporters to the user configuration.
-     *
-     * @param userConfig The user configuration to which the metric reporters will be added.
-     * @param injectCcMetricsReporter Flag indicating whether to inject the Cruise Control Metrics Reporter.
-     * @param injectKafkaJmxReporter  Inject the JMX Reporter into the configuration
-     * @param injectStrimziMetricsReporter Flag indicating whether to inject the Strimzi Metrics Reporter.
-     *
-     */
-    private void maybeAddMetricReporters(KafkaConfiguration userConfig, boolean injectCcMetricsReporter, boolean injectKafkaJmxReporter, boolean injectStrimziMetricsReporter) {
-        if (injectCcMetricsReporter) {
-            AbstractConfiguration.createOrAddListConfig(userConfig, "metric.reporters", CruiseControlMetricsReporter.CRUISE_CONTROL_METRIC_REPORTER);
-        }
-        if (injectKafkaJmxReporter) {
-            AbstractConfiguration.createOrAddListConfig(userConfig, "metric.reporters", "org.apache.kafka.common.metrics.JmxReporter");
+    private void printMetricReporters(AbstractConfiguration userConfig,
+                                      boolean injectCcMetricsReporter,
+                                      boolean injectKafkaJmxReporter,
+                                      boolean injectStrimziMetricsReporter) {
+        OrderedProperties props = new OrderedProperties();
+        String configKey = "metric.reporters";
+        // the JmxReporter is explicitly added when metric.reporters is not empty
+        List<String> reporterClasses = List.of(
+                CruiseControlMetricsReporter.CRUISE_CONTROL_METRIC_REPORTER,
+                "org.apache.kafka.common.metrics.JmxReporter",
+                StrimziMetricsReporterConfig.KAFKA_CLASS
+        );
+        List<Boolean> injectFlags = List.of(
+                injectCcMetricsReporter,
+                injectKafkaJmxReporter,
+                injectStrimziMetricsReporter
+        );
+        boolean hasUserMetricReporters = false;
+
+        for (int i = 0; i < reporterClasses.size(); i++) {
+            if (userConfig != null && !userConfig.getConfiguration().isEmpty() &&
+                    userConfig.getConfigOption(configKey) != null && injectFlags.get(i)) {
+                hasUserMetricReporters = true;
+                props.addPair(configKey, userConfig.getConfigOption(configKey));
+                userConfig.removeConfigOption(configKey);
```

Comment:
the above three lines of code doesn't use anything which is driving the for loop (i.e. `reporterClasses`, ...) why can't they be outside of the loop?

Reply:
These have to be executed only if there is user config AND an injection needed (i.e. we don't mess with user config unless we have to). I reworked the code to make it more readable and less error-prone.